### ISSUE TITLE
Choreographies should be able to figure out choice broadcasting

### ISSUE THREAD

**Issue by ashton314** [utahplt/chorex#1]
Consider this example from the paper:

``` haskell
if ℓ₁.𝑒 then
  ℓ₃.𝑒′ ↝ ℓ₂.𝑥;
  ℓ1 [L] ↝ ℓ2;
  ℓ₂.𝑥 + 2
else
  ℓ₃.𝑒′ ↝ ℓ₂.𝑥;
  ℓ1 [R] ↝ ℓ2;
  ℓ₂.𝑥 + 3
```

The paper goes on to say that only ℓ₃ never needs to be informed because its behavior is the same down both branches; not so for ℓ₂, since its behavior depends on what ℓ₁ decides at the test in the `if`.

We should be able to figure out what parties actually need to know what the decision is and automatically add the decision-sending code by looking at the branches and analyzing how actors' behaviors differ. So, I'd like to be able to write:

``` haskell
if ℓ₁.𝑒 then
  ℓ₃.𝑒′ ↝ ℓ₂.𝑥;
  ℓ₂.𝑥 + 2
else
  ℓ₃.𝑒′ ↝ ℓ₂.𝑥;
  ℓ₂.𝑥 + 3
```

Is this tantamount to asking, "how much can I lift out of the if?"

**Comment by ashton314** [comment]
Indeed, with commit d714874 I had to do some funny work around branches: whenever you enter an `if` expression, we switch into processing the branch blocks in a pseudo-CPS mode. This is because the decision communication needs to wrap the rest of the computation together. (See how endpoint projection is defined for 〚ℓ₁[d] ↝ ℓ₂; C〛.)

**Comment by bennn** [comment]
I'm confused. Why do branches need different processing?

Looking at this chunk from sec 5.3, the projection looks local. This translation could happen outside an `if`:
<img width="582" alt="Screenshot 2024-05-13 at 3 38 22 PM" src="https://github.com/utahplt/chorex/assets/1731829/94c59e24-39e0-43d3-9deb-5dfb3ff6bd4b">

Is there trouble implementing `choose` and `allow`?

**Comment by bennn** [comment]
HasChor does choice broadcasting automatically:
https://doi.org/10.1145/3607849

Sec 3.3 says it's for convenience, but may lead to more communication than necessary.

**Comment by ashton314** [comment]
As of e938bd0, `if` blocks broadcast choice everywhere by default. This can be manually slimmed down with the `notify: [Actor1, Actor2]` syntax.
